### PR TITLE
New Open APIs for case search on landing page

### PR DESCRIPTION
**Pull request by atulgupta2024**
Open APIs for case summary search by CNR and by other parameters like Case Number, Case type and year. 

## Summary by CodeRabbit

- **New Features**
	- Introduced a new OpenAPI specification for public APIs.
	- Added endpoints for retrieving case summaries by CNR number and by year, case type, and case number.
	- Added multiple new endpoints for case management, including search, verification, and existence checks.
	- Enhanced search capabilities with updated request and response schemas.
	- Standardized error handling across various endpoints for consistent messaging.
	- Expanded documentation for clearer guidance on API usage and parameters.
	- Implemented pagination and sorting options for case list retrieval.
	- Introduced utility classes for managing user data, workflows, and ID generation.
	- Added a README file for server setup and usage instructions.
	- Introduced a configuration file with essential application properties for database and Kafka settings.
	- Added new classes for handling case summaries and responses, enhancing data structure management.

### CODE REVIEW

File: api_specifications/open-api-0.1.0.yaml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+          content:
+            '*/*':
+              schema:
+                $ref: '#/components/schemas/CaseSummaryResponse'
+        '400':
+          description: querying case summary failed
+          content:
+            '*/*':
+              schema:
+                $ref: https://raw.githubusercontent.com/egovernments/DIGIT-OSS/master/core-services/docs/common-contract_v1-1.yml#/components/schemas/ErrorRes
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:45]: these are standard external references used across the project

File: api_specifications/open-api-0.1.0.yaml [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+          description: full name of the primary complainant
+        respondent:
+          type: string
+          description: full name of the primary respondent
+        advocateComplainant:
+          type: string
+          description: full name of the primary advocate of complainant
+        advocateRespondent:
+          type: string
+          description: full name of the primary advocate of respondent
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:166]: done

File: api_specifications/open-api-0.1.0.yaml [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          schema:
+            type: string
+            maxLength: 16
+            minLength: 16
+      responses:
+        '200':
+          description: ResponseInfo with case summary
+          content:
+            '*/*':
+              schema:
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:38]: done

**Comment by subhashini-egov** [comment]
@atulgupta2024 Some suggestions from coderabbit are really good. Please incorporate those.

File: api_specifications/open-api-0.1.0.yaml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        filingNumber:
+          type: string
+          maxLength: 14
+          minLength: 14
+          example: "KL-000075-2024"
+        filingDate:
+          type: string
+          format: date
+          description: filing date in YYYY-MM-DD format, in IST time zone
+          example: "2024-11-24"
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:169]: done

File: api_specifications/open-api-0.1.0.yaml [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+  version: 0.1.0
+tags:
+  - name: Open APIs
+paths:
+  /openapi/{tenantID}/case/cnr/{cnrNumber}:
+    get:
+      tags:
+        - Open APIs
+      summary: get case summary by cnr number
+      description: The API is used to get case summary details using the case CNR Number within a specific tenant.
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:14]: done

File: api_specifications/open-api-0.1.0.yaml [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            default: 10
+            minimum: 1
+            maximum: 100
+        - name: sort
+          in: query
+          description: Sorting criteria in the format `field,asc` or `field,desc`
+          required: false
+          schema:
+            type: string
+      responses:
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:114]: done

File: api_specifications/open-api-0.1.0.yaml [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            type: string
+            enum: [CMP, CC, ST]
+        - name: caseNumber
+          in: query
+          description: the case number (CMP or ST) of the case, whose summary is requested. This optional parameter. If not specified, system will return all CMP/CC/ST cases for the specified tenant and year
+          required: false
+          schema:
+            type: string
+            maxLength: 18
+            minLength: 9
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:90]: done

File: api_specifications/case-api-0.1.0.yaml [hunk 8]
```diff
@@ -778,7 +778,7 @@ components:
     CaseListResponse:
       type: object
       properties:
-        RequestInfo:
+        ResponseInfo:
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/case-api-0.1.0.yaml:781]: fixed

File: api_specifications/open-api-0.1.0.yaml [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+tags:
+  - name: Open APIs
+paths:
+  /openapi/{tenantID}/case/cnr/{cnrNumber}:
+    get:
+      tags:
+        - Open APIs
+      summary: get case summary by cnr number
+      description: The API is used to get case summary details using the case CNR Number within a specific tenant. 
+                  example - GET /openapi/KL/case/cnr/KLKM520008652024
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml]: ignore

File: api_specifications/open-api-0.1.0.yaml [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            minimum: 0
+        - name: limit
+          in: query
+          description: Number of items per page
+          required: false
+          schema:
+            type: integer
+            default: 10
+            minimum: 1
+            maximum: 100
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:107]: descriptions are as per pagination DIGIT component

File: api_specifications/open-api-0.1.0.yaml [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+          items:
+            $ref: '#/components/schemas/CaseSummary'
+        pagination:
+          $ref: >-
+            https://raw.githubusercontent.com/egovernments/DIGIT-OSS/master/core-services/docs/common-contract_v1-1.yml#/components/schemas/Pagination
+
+    CaseSummary:
+      type: object
+      description: this is a summary representation of CourtCase object. A limited set of fields, formatted appropriately are returned. Since the object already exists, there are no field level rules set for this summary object
+      properties:
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:154]: done

File: api_specifications/open-api-0.1.0.yaml [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+          required: false
+          schema:
+            type: string
+      responses:
+        '200':
+          description: ResponseInfo with case summary
+          content:
+            'application/json':
+              schema:
+                $ref: '#/components/schemas/CaseListSummaryResponse'
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:120]: added

File: api_specifications/open-api-0.1.0.yaml [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+          content:
+            'application/json':
+              schema:
+                $ref: '#/components/schemas/CaseSummaryResponse'
+        '400':
+          description: querying case summary failed
+          content:
+            'application/json':
+              schema:
+                $ref: https://raw.githubusercontent.com/egovernments/DIGIT-OSS/master/core-services/docs/common-contract_v1-1.yml#/components/schemas/ErrorRes
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:47]: ignore

File: api_specifications/open-api-0.1.0.yaml [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            properties:
+              statute:
+                type: string
+                example: "NIA"
+              sections:
+                type: array
+                items:
+                  type: string
+                example: ["S138", "S142(1)(a)", "S145"]
+          description: An array of statute and sections for the case
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:199]: Ignore. these are response fields. There values/formats are defined elsewhere.

**Comment by atulgupta2024** [comment]
@subhashini-egov all comments addressed. Some resolved, some ignored, because they are handled elsewhere.

**Comment by atulgupta2024** [comment]
@manimaarans ignore the open-api-0.1.0.yaml. The file - open-api-1.0.yaml is the final version

File: api_specifications/open-api-1.0.yaml [hunk 15]
```diff
@@ -0,0 +1,337 @@
+openapi: "3.1.0"
+info:
+  title: All Public/Open APIs supported by the platform
+  description: This API specification defines the various APIs that are open to external world. Anyone can call these. These are unauthenticated APIs. These need to managed via rate limiting settings at the gateway level to limit access
+  version: 1.0
+tags:
+  - name: Open APIs
+paths:
+  /openapi/v1/{tenantID}/case/cnr/{cnrNumber}:
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-1.0.yaml:9]: This end point will be consumed by a direct search with CNR. Refer to the FIGMA - https://www.figma.com/design/fMbBvJGlgA5hh4YD2bOF2I/On-court-Landing-Page?node-id=0-1&node-type=canvas&t=10j22VzC75fF1dbI-0

File: api_specifications/open-api-0.1.0.yaml [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+            maxLength: 2
+            minLength: 2
+        - name: year
+          in: path
+          description: if type= CMP, then year in which the case was registered. Can check based on registration date also. If type = CC/ST, then check against CourtCase.courtCaseNumber(year). The minimum year is set to 2024 as this is the year the system has gone live and the first case in the system is from 2024. No earlier cases exist.
+          required: true
+          schema:
+            type: integer
+            format: int32
+            minimum: 2024
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:85]: not required

File: api_specifications/open-api-0.1.0.yaml [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+            maxLength: 2
+            minLength: 2
+        - name: cnrNumber
+          in: path
+          description: the CNR number of the case in format SCDCECNNNNNNYYYY where SC=State Code, DC=District Code, EC=Establishment Code NNNNNN=Case Number and YYYY=Year, whose summary is requested
+          required: true
+          schema:
+            type: string
+            maxLength: 16
+            minLength: 16
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:33]: ignore

File: api_specifications/open-api-0.1.0.yaml [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+        caseId:
+          type: string
+          format: uuid
+          description: the primary key of the case. will be useful in further queries
+        caseNumber:
+          type: string
+          description: CMP/CC/ST number of the case based on the input query caseType. Sorted on case Number in reverse chronological order, i.e. the latest case at the top and earliest at the bottom
+          example: "CMP/1/2024"
+        caseTitle:
+          type: string
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml]: ignore

File: api_specifications/open-api-1.0.yaml [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+          content:
+            'application/json':
+              schema:
+                $ref: https://raw.githubusercontent.com/egovernments/DIGIT-OSS/master/core-services/docs/common-contract_v1-1.yml#/components/schemas/ErrorRes
+
+  /openapi/v1/{tenantID}/case/{year}/{caseType}:
+    get:
+      tags:
+        - Open APIs
+      summary: get case summary by case number for a specific year
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-1.0.yaml]: @manimaarans fixed

`summary: get a list of cases for a specific year based on specified case type`

File: api_specifications/open-api-1.0.yaml [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+            'application/json':
+              schema:
+                $ref: '#/components/schemas/CaseSummaryResponse'
+        '400':
+          description: querying case summary failed
+          content:
+            'application/json':
+              schema:
+                $ref: https://raw.githubusercontent.com/egovernments/DIGIT-OSS/master/core-services/docs/common-contract_v1-1.yml#/components/schemas/ErrorRes
+
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-1.0.yaml:60]: ignore

File: api_specifications/open-api-1.0.yaml [hunk 21]
```diff
@@ -0,0 +1,337 @@
+openapi: "3.0.0"
+info:
+  title: All Public/Open APIs supported by the platform
+  description: This API specification defines the various APIs that are open to external world. Anyone can call these. These are unauthenticated APIs. These need to managed via rate limiting settings at the gateway level to limit access
+  version: 0.1.0
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-1.0.yaml:5]: fixed

**Review by atulgupta2024 — COMMENTED** [review]
done

File: api_specifications/open-api-0.1.0.yaml [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+            minimum: 2024
+        - name: caseType
+          in: path
+          description: the type of the case CMP/CC/ST
+          required: true
+          schema:
+            type: string
+            enum: [CMP, CC, ST]
+        - name: caseNumber
+          in: query
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-0.1.0.yaml:178]: done

File: api_specifications/open-api-1.0.yaml [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+          format: int64
+          description: registration date in EPOCH time (milliseconds) format
+        nextHearingDate:
+          type: integer
+          format: int64
+          description: date of next scheduled hearing in EPOCH time (milliseconds) format
+        caseType:
+          type: string
+          description: fixed value based on case type
+          enum: ["Criminal Miscellaneous Petition", "Summary Trail"]
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-1.0.yaml:290]: CC case is not valid at this time. Removed reference of it

File: api_specifications/open-api-1.0.yaml [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+                example: "NIA"
+              sections:
+                type: array
+                items:
+                  type: string
+                example: ["S138", "S142(1)(a)", "S145"]
+          description: An array of statute and sections for the case
+        status:
+          type: string
+          description: CMP case not yet admitted, response will be Pending. ST Case already admitted, response will be Pending. ST case already registered, but queried using CMP, response will be Disposed.
```
[the remark this answers is not in the document]
Comment (atulgupta2024) [reply api_specifications/open-api-1.0.yaml:306]: description updated to 

CMP case not yet admitted, response will be Pending. ST Case already admitted, response will be Pending or Disposed if case is closed or outcome has been reached. ST case already registered, but queried using CMP, response will be Disposed. If queried by CNR, then pending or disposed will depend on the case status

**Pull request merged**